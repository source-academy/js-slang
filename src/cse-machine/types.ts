import * as es from 'estree'

import { Environment } from '../types'

export enum InstrType {
  RESET = 'Reset',
  WHILE = 'While',
  FOR = 'For',
  ASSIGNMENT = 'Assignment',
  UNARY_OP = 'UnaryOperation',
  BINARY_OP = 'BinaryOperation',
  POP = 'Pop',
  APPLICATION = 'Application',
  BRANCH = 'Branch',
  ENVIRONMENT = 'Environment',
  ARRAY_LITERAL = 'ArrayLiteral',
  ARRAY_ACCESS = 'ArrayAccess',
  ARRAY_ASSIGNMENT = 'ArrayAssignment',
  ARRAY_LENGTH = 'ArrayLength',
  MARKER = 'Marker',
  CONTINUE = 'Continue',
  CONTINUE_MARKER = 'ContinueMarker',
  BREAK = 'Break',
  BREAK_MARKER = 'BreakMarker'
}

interface BaseInstr {
  instrType: InstrType
  srcNode: es.Node
}

export interface WhileInstr extends BaseInstr {
  test: es.Expression
  body: es.Statement
}

export interface ForInstr extends BaseInstr {
  init: es.VariableDeclaration | es.Expression
  test: es.Expression
  update: es.Expression
  body: es.Statement
}

export interface AssmtInstr extends BaseInstr {
  symbol: string
  constant: boolean
  declaration: boolean
}

export interface UnOpInstr extends BaseInstr {
  symbol: es.UnaryOperator
}

export interface BinOpInstr extends BaseInstr {
  symbol: es.BinaryOperator
}

export interface AppInstr extends BaseInstr {
  numOfArgs: number
  srcNode: es.CallExpression
}

export interface BranchInstr extends BaseInstr {
  consequent: es.Expression | es.Statement
  alternate: es.Expression | es.Statement | null | undefined
}

export interface EnvInstr extends BaseInstr {
  env: Environment
}

export interface ArrLitInstr extends BaseInstr {
  arity: number
}

export type Instr =
  | BaseInstr
  | WhileInstr
  | AssmtInstr
  | AppInstr
  | BranchInstr
  | EnvInstr
  | ArrLitInstr

export type AgendaItem = es.Node | Instr

// Special class that cannot be found on the stash so is safe to be used
// as an indicator of a breakpoint from running the ECE machine
export class ECEBreak {}

// Special value that cannot be found on the stash so is safe to be used
// as an indicator of an error from running the ECE machine
export class ECError {
  constructor(public readonly error: any) {}
}
