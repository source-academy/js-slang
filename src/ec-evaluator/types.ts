import * as es from 'estree'

import { Context } from '..'
import { Environment } from '../types'
import { Agenda, Stash } from './interpreter'

export enum InstrType {
  RESET = 'Reset',
  WHILE = 'While',
  FOR = 'For',
  ASSIGNMENT = 'Assignment',
  UNARY_OP = 'UnaryOperation',
  BINARY_OP = 'BinaryOperation',
  POP = 'Pop',
  APPLICATION = 'Application',
  BRANCH = 'Branch',
  ENVIRONMENT = 'Environment',
  PUSH_UNDEFINED_IF_NEEDED = 'PushUndefinedIfNeeded',
  ARRAY_LITERAL = 'ArrayLiteral',
  ARRAY_ACCESS = 'ArrayAccess',
  ARRAY_ASSIGNMENT = 'ArrayAssignment',
  ARRAY_LENGTH = 'ArrayLength',
  MARKER = 'Marker',
  CONTINUE = 'Continue',
  CONTINUE_MARKER = 'ContinueMarker',
  BREAK = 'Break',
  BREAK_MARKER = 'BreakMarker'
}

interface BaseInstr {
  instrType: InstrType
}

export interface WhileInstr extends BaseInstr {
  test: es.Expression
  body: es.Statement
  srcNode: es.Node
}

export interface ForInstr extends BaseInstr {
  init: es.VariableDeclaration | es.Expression
  test: es.Expression
  update: es.Expression
  body: es.Statement
  srcNode: es.Node
}

export interface AssmtInstr extends BaseInstr {
  symbol: string
  constant: boolean
  declaration: boolean
  srcNode: es.Node
}

export interface UnOpInstr extends BaseInstr {
  symbol: es.UnaryOperator
  srcNode: es.Node
}

export interface BinOpInstr extends BaseInstr {
  symbol: es.BinaryOperator
  srcNode: es.Node
}

export interface AppInstr extends BaseInstr {
  numOfArgs: number
  srcNode: es.CallExpression
}

export interface BranchInstr extends BaseInstr {
  consequent: es.Expression | es.Statement
  alternate: es.Expression | es.Statement | null | undefined
  srcNode: es.Node
}

export interface EnvInstr extends BaseInstr {
  env: Environment
}

export interface ArrLitInstr extends BaseInstr {
  arity: number
}

export type Instr =
  | BaseInstr
  | WhileInstr
  | AssmtInstr
  | AppInstr
  | BranchInstr
  | EnvInstr
  | ArrLitInstr

export type AgendaItem = es.Node | Instr

export type CmdEvaluator = (
  command: AgendaItem,
  context: Context,
  agenda: Agenda,
  stash: Stash
) => void

// Special class that cannot be found on the stash so is safe to be used
// as an indicator of a breakpoint from running the ECE machine
export class ECEBreak {}

// Special value that cannot be found on the stash so is safe to be used
// as an indicator of an error from running the ECE machine
export class ECError {
  constructor(public readonly error: any) {}
}
