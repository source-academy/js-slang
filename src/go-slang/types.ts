export enum NodeType {
  ExpressionStatement = 'ExpressionStatement',
  Literal = 'Literal',
  UnaryExpression = 'UnaryExpression',
  BinaryExpression = 'BinaryExpression'
}

// NOTE: right now, the only type of source file is an expression statement
export type SourceFile = ExpressionStatement

type Expression = Literal | UnaryExpression | BinaryExpression

export interface Node {
  type: NodeType
}

export interface ExpressionStatement extends Node {
  type: NodeType.ExpressionStatement
  expression: Expression
}

export interface Literal extends Node {
  type: NodeType.Literal
  value: any
}

export type UnaryOperator = '+' | '-'

export interface UnaryExpression extends Node {
  type: NodeType.UnaryExpression
  operator: UnaryOperator
  argument: Expression
}

export type BinaryOperator =
  | '+'
  | '-'
  | '*'
  | '/'
  | '%'
  | '|'
  | '^'
  | '=='
  | '!='
  | '<'
  | '<='
  | '>'
  | '>='

export interface BinaryExpression extends Node {
  type: NodeType.BinaryExpression
  operator: BinaryOperator
  left: Expression
  right: Expression
}

export enum CommandType {
  UnaryOp = 'UnaryOp',
  BinaryOp = 'BinaryOp'
}

export interface Command {
  type: CommandType
}

export interface UnaryOp extends Command {
  type: CommandType.UnaryOp
  operator: UnaryOperator
}

export interface BinaryOp extends Command {
  type: CommandType.BinaryOp
  operator: BinaryOperator
}

export type Instruction = ExpressionStatement | Expression | UnaryOp | BinaryOp
