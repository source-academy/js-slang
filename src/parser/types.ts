import type { Program } from 'estree'

import type { Context, Chapter, Node, SourceError, Variant } from '../types'

export type { Options as AcornOptions } from 'acorn'
export type { ParserOptions as BabelOptions } from '@babel/parser'

export interface Parser<TOptions> {
  parse(
    programStr: string,
    context: Context,
    options?: Partial<TOptions>,
    throwOnError?: boolean
  ): Program | null
  validate(ast: Program, context: Context, throwOnError?: boolean): boolean
}

export interface Rule<T extends Node> {
  name: string
  disableFromChapter?: Chapter
  disableForVariants?: Variant[]
  checkers: {
    [name: string]: (node: T, ancestors: Node[]) => SourceError[]
  }
}
