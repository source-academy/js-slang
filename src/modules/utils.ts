/**
 * For tabs that are export default declarations, we need to remove
 * the `export default` bit from the front before they can be loaded
 * by js-slang
 */
export function evalRawTab(text: string) {
  if (text.startsWith('export default')) {
    text = text.substring(14)
  }

  return eval(text)
}
