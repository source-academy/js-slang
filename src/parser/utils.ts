import {
  Comment,
  ecmaVersion,
  Node,
  parse as acornParse,
  parseExpressionAt as acornParseAt,
  Position
} from 'acorn'
import { parse as acornLooseParse } from 'acorn-loose'
import { Program, SourceLocation } from 'estree'

import { Context } from '..'
import { DEFAULT_ECMA_VERSION } from '../constants'
import { SourceError } from '../types'
import { validateAndAnnotate } from '../validator/validator'
import { MissingSemicolonError, TrailingCommaError } from './errors'
import { AcornOptions, BabelOptions } from './types'

/**
 * Generates options object for acorn parser
 *
 * @param ecmaVersion ECMA version
 * @param errors error container
 * @param throwOnError throw on error if true else push to error container and resume exec
 * @param options partial acorn options
 * @returns
 */
export const createAcornParserOptions = (
  ecmaVersion: ecmaVersion,
  errors?: SourceError[],
  options?: Partial<AcornOptions>,
  throwOnError?: boolean
): AcornOptions => ({
  ecmaVersion,
  sourceType: 'module',
  locations: true,
  onInsertedSemicolon(_tokenEndPos: number, tokenPos: Position) {
    const error = new MissingSemicolonError(positionToSourceLocation(tokenPos, options?.sourceFile))
    if (throwOnError) throw error
    errors?.push(error)
  },
  onTrailingComma(_tokenEndPos: number, tokenPos: Position) {
    const error = new TrailingCommaError(positionToSourceLocation(tokenPos, options?.sourceFile))
    if (throwOnError) throw error
    errors?.push(error)
  },
  ...options
})

/**
 * Parses a single expression at a specified offset
 *
 * @param programStr program string
 * @param offset position offset
 * @param ecmaVersion ECMA version
 * @returns acorn AST Node if parse succeeds else null
 */
export function parseAt(
  programStr: string,
  offset: number,
  ecmaVersion: ecmaVersion = DEFAULT_ECMA_VERSION
): Node | null {
  try {
    return acornParseAt(programStr, offset, { ecmaVersion })
  } catch (_error) {
    return null
  }
}

/**
 * Parse a program, returning alongside comments found within that program
 *
 * @param programStr program string
 * @param ecmaVersion ECMA version
 * @returns tuple consisting of the parsed program, and a list of comments found within the program string
 */
export function parseWithComments(
  programStr: string,
  ecmaVersion: ecmaVersion = DEFAULT_ECMA_VERSION
): [Program, Comment[]] {
  let comments: acorn.Comment[] = []
  const acornOptions: AcornOptions = createAcornParserOptions(
    ecmaVersion,
    undefined,
    {
      onComment: comments
    },
    undefined
  )

  let ast: Program | undefined
  try {
    ast = acornParse(programStr, acornOptions) as unknown as Program
  } catch {
    comments = []
    ast = acornLooseParse(programStr, acornOptions)
  }

  return [ast, comments]
}

/**
 * Parse program with error-tolerant acorn parser
 *
 * @param programStr program string
 * @param context js-slang context
 * @returns ast for program string
 */
export function looseParse(programStr: string, context: Context): Program {
  return acornLooseParse(
    programStr,
    createAcornParserOptions(DEFAULT_ECMA_VERSION, context.errors)
  ) as unknown as Program
}

/**
 * TODO
 *
 * @param programStr program string
 * @param context js-slang context
 * @returns ast for program string
 */
export function typedParse(programStr: string, context: Context): Program {
  const ast: Program = looseParse(programStr, context)
  return validateAndAnnotate(ast, context)
}

/**
 * Converts acorn parser Position object to SourceLocation object
 *
 * @param position acorn Position object
 * @returns SourceLocation
 */
export const positionToSourceLocation = (position: Position, source?: string): SourceLocation => ({
  start: { ...position },
  end: { ...position, column: position.column + 1 },
  source
})

export const defaultBabelOptions: BabelOptions = {
  sourceType: 'module',
  plugins: ['typescript', 'estree']
}
