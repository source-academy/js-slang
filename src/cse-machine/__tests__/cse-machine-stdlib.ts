import { Chapter, Value, Variant } from '../../types'
import { stripIndent } from '../../utils/formatters'
import { expectResult, snapshotFailure } from '../../utils/testing'

test.each([
  [
    Chapter.SOURCE_1,
    `
    display('message');
    `,
    true,
    'message'
  ],

  [
    Chapter.SOURCE_1,
    `
    error('error!');
    `,
    false,
    undefined
  ],

  [
    Chapter.SOURCE_1,
    `
    is_undefined(undefined);
    `,
    true,
    true
  ],

  [
    Chapter.SOURCE_2,
    `
    is_undefined(null);
    `,
    true,
    false
  ],

  [
    Chapter.SOURCE_2,
    `
    is_null(undefined);
    `,
    true,
    false
  ],

  [
    Chapter.SOURCE_2,
    `
    is_null(null);
    `,
    true,
    true
  ],

  [
    Chapter.SOURCE_1,
    `
    is_string('string');
    `,
    true,
    true
  ],

  [
    Chapter.SOURCE_1,
    `
    is_string('true');
    `,
    true,
    true
  ],

  [
    Chapter.SOURCE_1,
    `
    is_string('1');
    `,
    true,
    true
  ],

  [
    Chapter.SOURCE_1,
    `
    is_string(true);
    `,
    true,
    false
  ],

  [
    Chapter.SOURCE_1,
    `
    is_string(1);
    `,
    true,
    false
  ],

  [
    Chapter.SOURCE_1,
    `
    is_number('string');
    `,
    true,
    false
  ],

  [
    Chapter.SOURCE_1,
    `
    is_number('true');
    `,
    true,
    false
  ],

  [
    Chapter.SOURCE_1,
    `
    is_number('1');
    `,
    true,
    false
  ],

  [
    Chapter.SOURCE_1,
    `
    is_number(true);
    `,
    true,
    false
  ],

  [
    Chapter.SOURCE_1,
    `
    is_number(1);
    `,
    true,
    true
  ],

  [
    Chapter.SOURCE_1,
    `
    is_boolean('string');
    `,
    true,
    false
  ],

  [
    Chapter.SOURCE_1,
    `
    is_boolean('true');
    `,
    true,
    false
  ],

  [
    Chapter.SOURCE_1,
    `
    is_boolean('1');
    `,
    true,
    false
  ],

  [
    Chapter.SOURCE_1,
    `
    is_boolean(true);
    `,
    true,
    true
  ],

  [
    Chapter.SOURCE_1,
    `
    is_boolean(1);
    `,
    true,
    false
  ],

  [
    Chapter.SOURCE_1,
    `
    is_function(display);
    `,
    true,
    true
  ],

  [
    Chapter.SOURCE_1,
    `
    is_function(x => x);
    `,
    true,
    true
  ],

  [
    Chapter.SOURCE_1,
    `
    function f(x) {
      return x;
    }
    is_function(f);
    `,
    true,
    true
  ],

  [
    Chapter.SOURCE_1,
    `
    is_function(1);
    `,
    true,
    false
  ],

  [
    Chapter.SOURCE_3,
    `
    is_array(1);
    `,
    true,
    false
  ],

  [
    Chapter.SOURCE_3,
    `
    is_array(pair(1, 2));
    `,
    true,
    true
  ],

  [
    Chapter.SOURCE_3,
    `
    is_array([1]);
    `,
    true,
    true
  ],

  [
    Chapter.SOURCE_3,
    `
    array_length([1]);
    `,
    true,
    1
  ],

  [
    Chapter.SOURCE_1,
    `
    parse_int('10', 10);
    `,
    true,
    10
  ],

  [
    Chapter.SOURCE_1,
    `
    parse_int('10', 2);
    `,
    true,
    2
  ],

  [
    Chapter.SOURCE_1,
    `
    is_number(get_time());
    `,
    true,
    true
  ],

  [
    Chapter.SOURCE_1,
    `
    const start = get_time();
    function repeatUntilDifferentTime() {
      if (start === get_time()) {
        return repeatUntilDifferentTime();
      } else {
        return true;
      }
    }
    repeatUntilDifferentTime();
    `,
    true,
    true
  ],

  [
    Chapter.SOURCE_2,
    `
    pair(1, 2);
    `,
    true,
    [1, 2]
  ],

  [
    Chapter.SOURCE_2,
    `
    list(1, 2);
    `,
    true,
    [1, [2, null]]
  ],

  [
    Chapter.SOURCE_2,
    `
    is_list(1);
    `,
    true,
    false
  ],

  [
    Chapter.SOURCE_2,
    `
    is_list(pair(1, 2));
    `,
    true,
    false
  ],

  [
    Chapter.SOURCE_2,
    `
    is_list(list(1, 2));
    `,
    true,
    true
  ],

  [
    Chapter.SOURCE_2,
    `
    head(pair(1, 2));
    `,
    true,
    1
  ],

  [
    Chapter.SOURCE_2,
    `
    tail(pair(1, 2));
    `,
    true,
    2
  ],

  [
    Chapter.SOURCE_2,
    `
    head(null);
    `,
    false,
    undefined
  ],

  [
    Chapter.SOURCE_2,
    `
    tail(null);
    `,
    false,
    undefined
  ],

  [
    Chapter.SOURCE_2,
    `
    head(1);
    `,
    false,
    undefined
  ],

  [
    Chapter.SOURCE_2,
    `
    tail(1);
    `,
    false,
    undefined
  ],

  [
    Chapter.SOURCE_2,
    `
    length(list(1, 2));
    `,
    true,
    2
  ],

  [
    Chapter.SOURCE_2,
    `
    length(1);
    `,
    false,
    undefined
  ]
] as [Chapter, string, boolean, Value][])(
  'Builtins work as expected %#',
  (chapter: Chapter, snippet: string, passing: boolean, returnValue: Value) => {
    if (passing) {
      return expectResult(stripIndent(snippet), {
        chapter,
        variant: Variant.EXPLICIT_CONTROL
      }).toEqual(returnValue)
    } else {
      return snapshotFailure(
        stripIndent(snippet),
        { chapter, variant: Variant.EXPLICIT_CONTROL },
        'fails'
      )
    }
  }
)
