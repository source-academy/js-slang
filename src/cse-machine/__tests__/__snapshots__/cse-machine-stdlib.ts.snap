// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Builtins work as expected 0: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "display('message');",
  "displayResult": Array [
    "\\"message\\"",
  ],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "message",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 1: fails 1`] = `
Object {
  "alertResult": Array [],
  "code": "error('error!');",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 1: Error: \\"error!\\"",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 2: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_undefined(undefined);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 3: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_undefined(null);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 4: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_null(undefined);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 5: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_null(null);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 6: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_string('string');",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 7: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_string('true');",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 8: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_string('1');",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 9: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_string(true);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 10: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_string(1);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 11: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_number('string');",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 12: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_number('true');",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 13: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_number('1');",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 14: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_number(true);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 15: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_number(1);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 16: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_boolean('string');",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 17: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_boolean('true');",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 18: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_boolean('1');",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 19: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_boolean(true);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 20: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_boolean(1);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 21: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_function(display);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 22: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_function(x => x);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 23: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function f(x) {
  return x;
}
is_function(f);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 24: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_function(1);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 25: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_array(1);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 26: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_array(pair(1, 2));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 27: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_array([1]);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 28: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "array_length([1]);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 1,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 29: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "parse_int('10', 10);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 10,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 30: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "parse_int('10', 2);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 31: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_number(get_time());",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 32: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const start = get_time();
function repeatUntilDifferentTime() {
  if (start === get_time()) {
    return repeatUntilDifferentTime();
  } else {
    return true;
  }
}
repeatUntilDifferentTime();",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 33: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "pair(1, 2);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": Array [
    1,
    2,
  ],
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 34: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "list(1, 2);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": Array [
    1,
    Array [
      2,
      null,
    ],
  ],
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 35: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_list(1);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 36: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_list(pair(1, 2));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 37: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_list(list(1, 2));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 38: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "head(pair(1, 2));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 1,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 39: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "tail(pair(1, 2));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 40: fails 1`] = `
Object {
  "alertResult": Array [],
  "code": "head(null);",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 1: Error: head(xs) expects a pair as argument xs, but encountered null",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 41: fails 1`] = `
Object {
  "alertResult": Array [],
  "code": "tail(null);",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 1: Error: tail(xs) expects a pair as argument xs, but encountered null",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 42: fails 1`] = `
Object {
  "alertResult": Array [],
  "code": "head(1);",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 1: Error: head(xs) expects a pair as argument xs, but encountered 1",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 43: fails 1`] = `
Object {
  "alertResult": Array [],
  "code": "tail(1);",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 1: Error: tail(xs) expects a pair as argument xs, but encountered 1",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 44: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "length(list(1, 2));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Builtins work as expected 45: fails 1`] = `
Object {
  "alertResult": Array [],
  "code": "length(1);",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 33: Error: tail(xs) expects a pair as argument xs, but encountered 1",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;
